engine/platform/win32/sys_win.c

```diff
@@ -46,6 +62,92 @@ qboolean Sys_DebuggerPresent( void )
 	return IsDebuggerPresent();
 }
 
+static void Sys_SetTimerHighResolution( void )
+{	
+	HMODULE hndl;
+	pfnNtSetTimerResolution_t pNtSetTimerResolution;
+	pfnNtQueryTimerResolution_t pNtQueryTimerResolution;
+	
+	hndl = GetModuleHandle( "ntdll.dll" );
+	pNtSetTimerResolution = (pfnNtSetTimerResolution_t)GetProcAddress( hndl, "NtSetTimerResolution" );
+	pNtQueryTimerResolution = (pfnNtQueryTimerResolution_t)GetProcAddress( hndl, "NtQueryTimerResolution" );
+
+	if( pNtSetTimerResolution && pNtQueryTimerResolution )
+	{
+		// undocumented NT API functions, for some mystery reason works better than timeBeginPeriod
+		ULONG min, max, cur;
+		pNtQueryTimerResolution(&min, &max, &cur);
+		pNtSetTimerResolution(max, 1, &cur);
+	}
+	else
+	{
+		// more conventional method to set timer resolution, using as fallback
+		TIMECAPS tc;
+		timeGetDevCaps( &tc, sizeof( TIMECAPS ));
+		timeBeginPeriod( tc.wPeriodMin );
+	}
+}
+
+static DWORD WINAPI Sys_TimerThread( LPVOID a )
+{
+	double oldtime = 0;
+	LARGE_INTEGER delay;
+	while (true)
+	{
+		WaitForSingleObject( g_timer.hMutex, INFINITE ); // lock mutex
+		double realtime = Platform_DoubleTime();
+		double delta = g_timer.flInterval - ( realtime - oldtime );
+		oldtime = realtime;
+
+		if (delta > 0)
+			delta = 0;
+
+		delay.QuadPart = -1 * 1e7 * ( g_timer.flInterval + delta );
+		ReleaseMutex( g_timer.hMutex ); // unlock mutex
+
+		SetWaitableTimer( g_timer.hTimer, &delay, 0, NULL, NULL, 0 );
+		WaitForSingleObject( g_timer.hTimer, INFINITE );
+		SetEvent( g_timer.hEvent );
+	}
+}
+
+void Platform_Delay( double time )
+{
+	WaitForSingleObject( g_timer.hMutex, INFINITE ); // lock mutex
+	g_timer.flInterval = time;
+	ReleaseMutex( g_timer.hMutex ); // unlock mutex
+	WaitForSingleObject( g_timer.hEvent, INFINITE ); // wait when event will be triggered
+}
+
+void Platform_TimerInit( void )
+{
+	Sys_SetTimerHighResolution();
+	g_timer.hTimer = CreateWaitableTimer(NULL, TRUE, NULL);
+	g_timer.hEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
+	g_timer.flInterval = 1.0 / 60.0;
+
+	g_timer.hMutex = CreateMutex(
+		NULL,              // default security attributes
+		FALSE,             // initially not owned
+		NULL);             // unnamed mutex
+
+	g_timer.hThread = CreateThread(
+		NULL,                   // default security attributes
+		0,                      // use default stack size
+		Sys_TimerThread,		// thread function name
+		0,						// argument to thread function
+		0,                      // use default creation flags
+		&g_timer.iThreadID);
+}
+
+void Platform_TimerShutdown( void )
+{
+	CloseHandle( g_timer.hThread );
```

Comment:
You can close thread handle when you create thread.
One correct way to stop thread - return error code from thread.
CloseHandle just clears thread handle and nothing more.

Reply:
Should I stop thread in this case at all?

Reply:
You can stop thread via [ExitThread](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitthread) and [TerminateThread](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminatethread) callbacks, but it's not recommended:

> TerminateThread is used to cause a thread to exit. When this occurs, the target thread has no chance to execute any user-mode code. DLLs attached to the thread are not notified that the thread is terminating. The system frees the thread's initial stack.

> Windows Server 2003 and Windows XP:  The target thread's initial stack is not freed, causing a resource leak.

> TerminateThread is a dangerous function that should only be used in the most extreme cases. You should call TerminateThread only if you know exactly what the target thread is doing, and you control all of the code that the target thread could possibly be running at the time of the termination. For example, TerminateThread can result in the following problems:

>    If the target thread owns a critical section, the critical section will not be released.
>    If the target thread is allocating memory from the heap, the heap lock will not be released.
>    If the target thread is executing certain kernel32 calls when it is terminated, the kernel32 state for the thread's process could be inconsistent.
 >   If the target thread is manipulating the global state of a shared DLL, the state of the DLL could be destroyed, affecting other users of the DLL.

Or you can save thread handle and reuse it later.

Reply:
So I think it's better to stop thread by setting flag in main thread and checking it in timer thread, instead of all this nightmare xD

Reply:
Anyway, you don't need to save thread handle and probably no reason to clear all other open handles if process will be stopped.